File: 08_AndroidBasic/SampleThread2/app/src/main/java/org/techtown/thread/MainActivity.java
package org.techtown.thread;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

/* UI 를 처리하는 스레드를 메인 스레드라 부른다. 메인 스레드에서 이미 UI에 접근하고 있으므로
  새로 생성한 스레드에서는 핸들러 객체를 이용해서 메시지를 전달함으로서 메인스레드에서 처리하도록 만들 수 있다.

  다음은 안드로이드에서 제공하는 두가지 시나리오다.
  지연시간이 길어질 수 있는 앱이라면 오랜 시간 작업을 수행하는 코드를 별도로 분리한 다음 UI에 응답을 보내는 방식으로 개발한다.

  1. 서비스 사용하기 : 백그라운드 작업은 서비스로 실행하고 사용자에게는 알림 서비스로 알려준다. 만약 엑티비티로 결과 값을
  전달하고 이를 이용해 다른 작업을 수행한다면 브로드캐스팅으로 값을 전달한다.
  2. 스레드 사용하기 : 스레드는 같은 프로세스 안에서 동작하므로 작업 수행의 결과를 바로 처리 가능하다. 그러나 UI 객체는
  직접 접근할 수 없으므로 핸들러를 사용한다.
  */
public class MainActivity extends AppCompatActivity {
    TextView textView;

    MainHandler handler;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        textView = findViewById(R.id.textView);

        Button button = findViewById(R.id.button);
        button.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                BackgroundThread thread = new BackgroundThread();
                thread.start();
            }
        });

        handler = new MainHandler();
    }

    // 기본 접근 같은 클래스, 패키지 이용
    class BackgroundThread extends Thread{
        int value = 0;

        public void run() {
            for(int i = 0 ; i < 100 ; i++){
                try{
                    Thread.sleep(1000);
                }catch (Exception e){

                }
                value += 1;
                Log.d("Thread", "value : " + value);
                // 메인스레드가 아닌곳에서 UI 객체에 접근할 수 없다. 에러코드는 안뜨지만 버튼을 터치하는 순간 에러가 발생한다.'
                // 핸들러를 이용해야한다.
                // textView.setText("value 값 : " + value);

                // 핸들러가 관리하는 메시지 큐에서 처리할 수 있는 메시지 객체를 반환받는다.
                Message message = handler.obtainMessage();
                Bundle bundle = new Bundle();
                bundle.putInt("value", value);
                message.setData(bundle);

                // 셋팅한 메시지 객체를 핸들러로 보낸다.
                handler.sendMessage(message);
            }
        }
    }

    /* 앱을 실행할 때 프로세스가 만들어지면 그와 함께 메인 스레드도 생성된다. 최상위에서 관리되는 앱 구성 요소인 액티비티, 브로드캐스트 수신자 등
    새로 만들어지는 윈도우를 관리하기 위한 메시지 큐를 실행한다. 메시지큐로 순차적으로 코드를 진행하는데, 메인스레드에서 이 처리할 메시지를 전달하는 역할을 핸들러가 수행한다.
    실행하려는 특정 기능이 있을때 핸들러가 포함된 스레드에서 순차적으로 실행시킬때 사용된다. 핸들러를 이용해 미래에 어떤 시점에 이용되도록 스케줄링도 가능하다. */

    /*
     * 핸들러가 관리하는 메시지 큐에서 처리할 수 있는 메시지 객체 하나를 참조하기 위해 obtainMessage 메서드를 이용하고 호출의 결과로 메시지 객체를 반환받는다.
     * 이 메시지 객체에 필요한 정보를 넣은 후 sendMessage 메서드를 이용해 메시지 큐에 넣을 수 있다. 메시지 큐에 들어간 메시지는 순서대로 핸들러가 처리하게 되며
     * 이때 handleMessage 메서드에 정의한 기능이 수행된다. 이때 handleMessage에 들어 있는 코드가 수행되는 위치는 새로 만든 스레드가 아닌 메인 스레드이다.
     * */
    class MainHandler extends Handler {

        // 핸들러 안에서 전달받은 메시지 객체 처리하기
        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);

            Bundle bundle = msg.getData();
            int value = bundle.getInt("value");
            textView.setText("value 값 : " + value);
        }
    }
}

File: 05. Android/app/src/main/java/com/bignerdranch/android/geoguiz/MainActivity.kt
package com.bignerdranch.android.geoguiz

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider

// 최상위 수준 속성
// 특정 클래스의 인스턴스를 생성하지 않고 앱이 실행되는 동안 계속 보존
private const val TAG = "MainActivity"

// Bundle 객체에 저장될 데이터의 키
private const val KEY_INDEX = "index"

// 자식 액티비티로 돌려받을 데이터
private const val REQUEST_CODE_CHEAT = 0

class MainActivity : AppCompatActivity() {
    private lateinit var trueButton : Button
    private lateinit var falseButton : Button
    private lateinit var previousButton : ImageButton
    private lateinit var nextButton : ImageButton
    private lateinit var questionTextView : TextView
    private lateinit var cheatButton : Button

    // ViewModelProvider(this).get(QuizViewModel::class.java)
    //        val provider : ViewModelProvider = ViewModelProvider(this)
    //        val quizViewModel = provider.get(QuizViewModel::class.java)
    //        Log.d(TAG, "Got a QuizViewModel: $quizViewModel")

    // quizViewModel을 var이 아닌 val 속성으로 선언이 가능하다. ViewModel은 인스턴스 생성시 한번만 저장하기 때문에 lazy를 사용한다.
    // 최초 사용시에만 해당 인스턴스를 생성한다. 초기화를 늦출수 있음.
    private val quizViewModel : QuizViewModel by lazy{
        ViewModelProvider(this).get(QuizViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // public static Int d(String tag, String msg) 이용
        // ERROR Log.e() 에러
        // WARNING Log.w() 경고
        // INFO Log.i() 정보성 메시지
        // DEBUG Log.d() 디버깅 출력
        // VERBOSE Log.v() 개발 전용
        Log.d(TAG, "onCreate(Bundle?) called");
        setContentView(R.layout.activity_main);

        // 값이 있다면 저장. 없거나 null 이면 0으로 저장
        // 인스턴스 최초 생성시 null
        val currentIndex = savedInstanceState?.getInt(KEY_INDEX,0) ?: 0;
        quizViewModel.currentIndex = currentIndex;

        trueButton = findViewById(R.id.true_button);
        falseButton = findViewById(R.id.false_button);
        previousButton = findViewById(R.id.previous_button);
        nextButton = findViewById(R.id.next_button);
        questionTextView = findViewById(R.id.question_text_view);
        cheatButton = findViewById(R.id.cheat_button);

        // True 버튼
        trueButton.setOnClickListener{ view : View ->
            checkAnswer(true);
        }
        
        // False 버튼
        falseButton.setOnClickListener{ view : View ->
            checkAnswer(false);
        }

        // Previous 버튼
        previousButton.setOnClickListener{
            quizViewModel.moveToPrevious()
            updateQuestion()
        }

        // Next 버튼
        nextButton.setOnClickListener{
            quizViewModel.moveToNext()
            updateQuestion()
        }

        cheatButton.setOnClickListener{
            // CheatActivity를 시작시킨다.
//            var intent = Intent(this, CheatActivity::class.java)
            val answerIsTrue = quizViewModel.currentQuestionAnswer
            val intent = CheatActivity.newIntent(this@MainActivity, answerIsTrue)
            // 액티비티 전환을 위해 사용
//            startActivity(intent)
            // 자식 액티비티로부터 리턴값을 전달받기 위해 사용
            startActivityForResult(intent, REQUEST_CODE_CHEAT)
        }

        updateQuestion()
    }

    override fun onStart(){
        super.onStart()
        Log.d(TAG, "onStart() called")
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "onResume() called")
    }

    override fun onPause() {
        super.onPause()
        Log.d(TAG, "onPause() called")
    }

    // ViewModel, SIS
    /** saveInstanceState
     *  프로세스 종료, 장치의 구성이 변경에도 Bundle 객체를 사용해 엑티비티 레코드를 저장한다.
     *  엑티비티 최초 실행시 Bundel 객체 참조는 null. 화면 전환 등 구성 변경시 onSaveInstanceState(Bundle)을 호출 보존 데이터를 활용한다.
     *  이후 생성된 엑티비티 인스턴스에도 안드로이드 OS가 저장된 Bundle 상태를 onCreate(Bundle?)의 인자로 전달한다.
     **/
    /** ViewModel
     *  ViewModel은 엑티비티의 동적 데이터를 처리할 때 편리하다.
     *  구성 변경이 있더라도 다운로드 작업을 계속하게 해준다. 사용자가 엑티비티 종료시 자동으로 클린업된다.
     *  프로세스 종료시에는 ViewModel이 클린업을 처리하지 못하며 자신이 가진 프로세스와 함께 메모리에서 제거된다.
     *  SIS가 주목받는 이유가 이때문이다. 허나 제약이 존재한다. SIS는 직렬화되어 디스크에 저장되므로 크거나 복잡한 객체는 피해야한다.
     *  !) lifecycle-viewmodel-savedstate 라이브러리 배포로 인해 해당 단점 또한 완화 가능. 두 가지를 적절히 사용하면 좋다.
     *  소량의 정보는 SIS, 많은 데이터 저장 및 캐시 저장을 위해서는 ViewModel을 사용한다.
     **/
    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        super.onSaveInstanceState(savedInstanceState)
        Log.d(TAG, "onSaveInstanceState")
        savedInstanceState.putInt(KEY_INDEX, quizViewModel.currentIndex);
    }

    override fun onStop() {
        super.onStop()
        Log.d(TAG, "onStop() called")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy() called")
    }

    private fun updateQuestion(){
//        Log.d(TAG, "Updating question text", Exception())
        val questionTextResId = quizViewModel.currentQuestionText
        questionTextView.setText(questionTextResId)
    }

    // 뷰로 부터 받은 값 확인
    private fun checkAnswer(userAnswer : Boolean){
        val correctAnswer = quizViewModel.currentQuestionAnswer
        if (userAnswer == correctAnswer){
            Toast.makeText(this, R.string.correct_toast, Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(this, R.string.incorrect_toast, Toast.LENGTH_SHORT).show();
        }
    }
}